src/QPSolver.cpp

```diff
@@ -214,6 +224,29 @@ void QPSolver::nrVars(const std::vector<rbd::MultiBody>& mbs,
 		c->updateNrVars(mbs, data_);
 	}
 
+	if(dependencies.size())
+	{
+		std::vector<int> fullToReduced(data_.nrVars_, -1);
+		std::vector<int> reducedToFull(data_.nrVars_ - static_cast<int>(dependencies.size()), -1);
+		std::vector<int> removedVars; removedVars.reserve(dependencies.size() + 1);
+		for(const auto & d : dependencies)
+		{
+			removedVars.push_back(std::get<1>(d));
+		}
+		removedVars.push_back(data_.nrVars_);
+		std::sort(removedVars.begin(), removedVars.end());
+		size_t shift = 0;
+		for(size_t i = 0; i < fullToReduced.size(); ++i)
+		{
+			if(i == static_cast<size_t>(removedVars[shift])) { shift++; continue; }
+			fullToReduced[i] = static_cast<int>(i - shift);
+			reducedToFull[i-shift] = static_cast<int>(i);
+		}
+		solver_->setReductionParameters(std::move(fullToReduced),
+																		std::move(reducedToFull),
+																		std::move(dependencies));
+	}
+
```

Comment:
The logic of this block is kind of obscure... Moreover, as the whole algorithm relies on `fullToReduced`, `reducedToFull` and `dependencies` to have compatible sizes, wouldn't it be better to perform this transformation in `setReductionParameters`? We would thus:
- Avoid potential crashes from users incorrectly setting `fullToReduced`
- Hide this complexity behind a specific method instead of overstuffing `nrVars`
- Keep the without mimic logic better encapsulated
- Maybe even make `setReductionParameters` protected if the user is not supposed to call it manually.

For the block itself:
- In general, add the explanation that you are looking for a "compression" of the whole variables into the reduced set, i.e. you want to remove all the mimics while keeping the same order for the other variables.
- Explain why you push `data_.nrVars_` to `removedVars` (I guess to continue after having found the last mimic without crashing)
- I'd prefer that you iterate over `nrVars_` rather that `fullToReduced.size()` but I guess that poses `unsigned int` to `int` cast problems
- I'd add a comment saying that `shift` represents the current number of mimic joints encountered so far.